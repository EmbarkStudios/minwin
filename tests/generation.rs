// use tracing_subscriber::{prelude::__tracing_subscriber_SubscriberExt, Layer};
//     let layer = tracing_subscriber::fmt::layer()
//         .with_test_writer()
//         .with_filter(tracing::metadata::LevelFilter::from_level(
//             tracing::metadata::Level::TRACE,
//         ));
//     let reg = tracing_subscriber::Registry::default().with(layer);
//     let dis = tracing::Dispatch::new(reg);
//     //let _guard = tracing::dispatcher::set_global_default(dis);

use anyhow::Context as _;

fn generate(input: &str, format: bool, convert_case: bool) {
    let mut parser = minwin::Parser::default();
    parser.add_file(format!("tests/data/{input}.rs"));
    let mut parsed = parser.parse().pop().unwrap();

    let mut hints = minwin::Hints::default();
    parsed.gather_hints(&mut hints);

    let md = minwin::MetadataFiles::new().expect("failed to gather metadata files");
    let resolved = minwin::Resolver::flatten(&md, hints).expect("failed to resolve metadata");

    let generated = parsed
        .iter_bind_modules()
        .enumerate()
        .map(|(i, m)| {
            let ts = minwin::generate(&resolved, m, convert_case)
                .with_context(|| format!("{}", m.ident))?;
            Ok((i, ts))
        })
        .collect::<anyhow::Result<Vec<_>>>()
        .expect("failed to generate modules");

    for (index, ts) in generated {
        parsed
            .replace_module(index, ts)
            .expect("failed to replace {index}");
    }

    let generated = parsed.generate(format).expect("failed to generate");

    insta::assert_snapshot!(format!("{input}_{format}"), generated);
}

/// Ensures we can generate function pointers for regular functions for use
/// with eg. `GetProcAddress`
#[test]
fn delegates_for_functions() {
    generate("parking_lot", true, false);
}

/// Ensures we can generate more complex types
/// 1. Arch specific structs
/// 2. Records with alignment
/// 3. Records with packing
/// 4. Nested structs/unions
#[test]
fn complex() {
    generate("md_writer", true, false);
}

/// While the windows metadata professes to encourage enums instead of plain
/// integer constants, there...aren't that many. But just check we can emit the
/// basics
#[test]
fn enums() {
    generate("enums", true, false);
}

/// Ensures we can emit function pointers for actual function pointers
#[test]
fn delegates() {
    generate("delegates", true, false);
}
