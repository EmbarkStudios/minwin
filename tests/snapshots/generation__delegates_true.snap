---
source: tests/generation.rs
expression: generated
---
#[minwin]
mod bindings {
    pub type BOOL = i32;
    pub type HANDLE = isize;
    #[repr(C)]
    pub struct OVERLAPPED_0_0 {
        pub Offset: u32,
        pub OffsetHigh: u32,
    }
    #[repr(C)]
    pub union OVERLAPPED_0 {
        pub Anonymous: std::mem::ManuallyDrop<OVERLAPPED_0_0>,
        pub Pointer: *mut std::ffi::c_void,
    }
    #[repr(C)]
    pub struct OVERLAPPED {
        pub Internal: usize,
        pub InternalHigh: usize,
        pub Anonymous: OVERLAPPED_0,
        pub hEvent: HANDLE,
    }
    pub type LPOVERLAPPED_COMPLETION_ROUTINE = Option<
        unsafe extern "system" fn(
            dwErrorCode: u32,
            dwNumberOfBytesTransfered: u32,
            lpOverlapped: *mut OVERLAPPED,
        ),
    >;
    #[link(name = "kernel32")]
    extern "system" {
        pub fn BindIoCompletionCallback(
            FileHandle: HANDLE,
            Function: LPOVERLAPPED_COMPLETION_ROUTINE,
            Flags: u32,
        ) -> BOOL;
    }
}

